src/logging/timestamp_checkpoint_manager.cpp

```diff
@@ -0,0 +1,1084 @@
+          input_buffer, schema->GetType(column_id), NULL);
+      tuple->SetValue(column_id, value, pool.get());
+    }
+
+    // insert tuple into the table without foreign key check to avoid an error
+    // which occurs in tables with the mutual foreign keys each other
+    ItemPointer location =
+        table->InsertTuple(tuple.get(), txn, &index_entry_ptr, false);
+    if (location.block != INVALID_OID) {
+      concurrency::TransactionManagerFactory::GetInstance().PerformInsert(
+          txn, location, index_entry_ptr);
+      insert_tuple_count++;
+    } else {
+      LOG_ERROR("Tuple insert error for table %d", table->GetOid());
+    }
+  }
+
+  return insert_tuple_count;
+}
+
+oid_t TimestampCheckpointManager::RecoverTableDataWithDuplicateCheck(
+    storage::DataTable *table, FileHandle &file_handle,
+    concurrency::TransactionContext *txn) {
+	size_t table_size = LoggingUtil::GetFileSize(file_handle);
+  if (table_size == 0) return 0;
+  std::unique_ptr<char[]> data(new char[table_size]);
+  if (LoggingUtil::ReadNBytesFromFile(file_handle, data.get(), table_size) == false) {
+    LOG_ERROR("Checkpoint table file read error");
+    return 0;
+  }
+  CopySerializeInput input_buffer(data.get(), table_size);
+
+  LOG_TRACE("Recover table %d data with duplicate check (%lu byte)",
+            table->GetOid(), table_size);
+
+  // look for all primary key columns
+  std::vector<oid_t> pk_columns;
+  auto schema = table->GetSchema();
+  oid_t column_count = schema->GetColumnCount();
+  for (oid_t column_id = 0; column_id < column_count; column_id++) {
+    if (schema->GetColumn(column_id).IsPrimary()) {
+      pk_columns.push_back(column_id);
+    }
+  }
+
+  // recover table tuples
+	std::unique_ptr<type::AbstractPool> pool(new type::EphemeralPool());
+  oid_t insert_tuple_count = 0;
+  while (input_buffer.RestSize() > 0) {
+    // recover values on each column
+    std::unique_ptr<storage::Tuple> tuple(new storage::Tuple(schema, true));
+    ItemPointer *index_entry_ptr = nullptr;
+    for (oid_t column_id = 0; column_id < column_count; column_id++) {
+      auto value = type::Value::DeserializeFrom(
+          input_buffer, schema->GetType(column_id), NULL);
+      tuple->SetValue(column_id, value, pool.get());
+    }
+
+    // duplicate check
+    // if all primary key values are existed, the tuple is not stored in the
+    // table
+    bool duplicated = false;
+    for (oid_t tg_offset = 0; tg_offset < table->GetTileGroupCount();
+         tg_offset++) {
+      auto tile_group = table->GetTileGroup(tg_offset);
+      auto max_tuple_count = tile_group->GetNextTupleSlot();
+      for (oid_t tuple_id = 0; tuple_id < max_tuple_count; tuple_id++) {
+        // check all primary key columns
+        bool check_all_pk_values_same = true;
+        for (auto pk_column : pk_columns) {
+          if (tile_group->GetValue(tuple_id, pk_column)
+                  .CompareNotEquals(tuple->GetValue(pk_column)) ==
```

Comment:
Should the duplicate check iterates tuples in the current table and compare the tuple recovered from the file with each of them for primary key columns? The tuple recovered shouldn't be stored in the table as long as there is one current tuple whose pk columns match its corresponding ones. It doesn't necessarily have to be no current tuples mismatching.

Reply:
To reduce processes of the check, if a value in the existed table's tuple in a primary key column differs from a recovered value, then the loop of the pk loop would break. This means it finds these tuples mismatching, and reduces the check processes of the rest of primary keys.